code/tests/generic_322_2.cpp

```diff
@@ -0,0 +1,150 @@
+/*
+Reproducing xfstest generic/322 _write_after_fsync_rename_test
+
+1. Create file foo and write some contents into it and do fsync
+2. Write some more contents and do sync again
+3. Rename foo to bar and do fsync
+
+After a crash at random point, bar should be present and should have the same contents
+written to foo.
+
+https://github.com/kdave/xfstests/blob/master/tests/generic/322
+*/
+
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <string>
+#include <iostream>
+#include <dirent.h>
+#include <cstring>
+#include <errno.h>
+
+#include "BaseTestCase.h"
+#include "../user_tools/api/workload.h"
+#include "../user_tools/api/actions.h"
+#define TEST_FILE_FOO "foo"
+#define TEST_FILE_BAR "bar"
+#define TEST_DIR_A "test_dir_a"
+
+
+using fs_testing::tests::DataTestResult;
+using fs_testing::user_tools::api::WriteData;
+using fs_testing::user_tools::api::WriteDataMmap;
+using fs_testing::user_tools::api::Checkpoint;
+using std::string;
+
+#define TEST_FILE_PERMS  ((mode_t) (S_IRWXU | S_IRWXG | S_IRWXO))
+
+namespace fs_testing {
+namespace tests {
+
+
+class Generic322_2: public BaseTestCase {
+ public:
+  virtual int setup() override {
+
+    // Create test directory A.
+	string dir_path = mnt_dir_ + "/" TEST_DIR_A;
+	int res = mkdir(dir_path.c_str(), 0777);
+    if (res < 0) {
+      return -1;
+    }
+
+    //Create file foo in TEST_DIR_A 
+    const int fd_foo = open(foo_path.c_str(), O_RDWR | O_CREAT, TEST_FILE_PERMS);
+    if (fd_foo < 0) {
+      return -1;
+    }
+
+    // Write some contents to the file
+    const char *buf = file_contents.c_str();
+    if (pwrite(fd_foo, buf, strlen(buf), 0) < 0) {
+    	return -2;
+    }
+
+    // Sync the file
+    if (fsync(fd_foo) < 0) {
+    	return -1;
+    }
+
+    // write more contents
+    const char *buf2 = file_contents2.c_str();
+    if (pwrite(fd_foo, buf2, strlen(buf2), 0) < 0) {
+    	return -2;
+    }
+
+    // sync again
+    sync();
+    close(fd_foo);
+    return 0;
+  }
+
+  virtual int run() override {
+
+	// Rename the foo to bar
+	if (rename(foo_path.c_str(), bar_path.c_str()) != 0) {
```

Comment:
This test is no different from the one above(generic_322). See that the run() method of both these tests have the same workload - so it wouldn't make a difference if you rename a file of size 4K vs a file of size 8K (I assume the behaviour will be the same). If instead you had one of the writes to the file after the rename in run() for example, you can check that the contents you wrote after a rename are persisted or not.

Reply:
Yes. But xfstests had two separate tests -https://github.com/kdave/xfstests/blob/master/tests/generic/322

Reply:
Note that there's a fsync after the first write, but after the second write they do a sync_range (which maps to http://man7.org/linux/man-pages/man2/sync_file_range.2.html I guess.) There is no fsync after the second write and they rename the file therafter. The bug report says btrfs dropped the extents between two fsync (one after the first write and the second after rename). However in this CM test, the sync after write-2 will persist it anyway before rename. To replicate this test, I'd move the second write to the run() method, perform a sync_file_range for the offsets modified, and then rename . Makes sense?

Reply:
Fixed.